### PR TITLE
feat(secret): add line from dockerfile where secret was added to secret result

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
For images: secrets contain the string from `dockerfile` where this secret was added from.
Also, this line is contained in the table report.
e.g.
```bash
/secret.txt (secrets)

Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 1)

CRITICAL: AWS (aws-access-key-id)
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
AWS Access Key ID
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 /secret.txt:1
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 This secret is added in 'echo "AWS_ACCESS_KEY_ID=AKIA0123456789ABCDEF" > secret.txt'
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1 [ AWS_ACCESS_KEY_ID=********************
   2   
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


```

## Related PRs
- [x] #2790

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/fanal/artifact/image/image.go [hunk 1]
```diff
@@ -133,45 +138,65 @@ func (Artifact) Clean(_ types.ArtifactReference) error {
...
 	// Pass an empty config scanner option so that the cache key can be the same, even when policies are updated.
 	imageKey, err := cache.CalcKey(imageID, a.analyzer.ImageConfigAnalyzerVersions(), nil, artifact.Option{})
 	if err != nil {
 		return "", nil, nil, err
 	}
 
-	layerKeyMap := map[string]string{}
+	var createdBy []string
+	// save createdBy fields in order of layers
+	for i := 0; i < len(configFile.History); i++ {
```
Comment (knqyf263) [inline pkg/fanal/artifact/image/image.go:150]:
```suggestion
	for i, history := range configFile.History {
```

File: pkg/fanal/artifact/image/image.go [hunk 2]
```diff
@@ -133,45 +138,65 @@ func (Artifact) Clean(_ types.ArtifactReference) error {
...
+	for i, diffID := range diffIDs {
 		blobKey, err := cache.CalcKey(diffID, a.analyzer.AnalyzerVersions(), hookVersions, a.artifactOption)
 		if err != nil {
 			return "", nil, nil, err
 		}
 		layerKeys = append(layerKeys, blobKey)
-		layerKeyMap[blobKey] = diffID
+
+		c := ""
+		if len(createdBy) > 0 {
```
Comment (knqyf263) [inline pkg/fanal/artifact/image/image.go:171]:
```suggestion
		if len(createdBy) > i {
```

File: pkg/fanal/artifact/image/image.go [hunk 3]
```diff
@@ -133,45 +138,65 @@ func (Artifact) Clean(_ types.ArtifactReference) error {
...
+		}
+		c := strings.TrimPrefix(strings.TrimPrefix(h.CreatedBy, "/bin/sh -c "), "#(nop) ")
+		createdBy = append(createdBy, c)
+	}
+
+	layerKeyMap := map[string]LayerInfo{}
 	hookVersions := a.handlerManager.Versions()
 	var layerKeys []string
-	for _, diffID := range diffIDs {
+	for i, diffID := range diffIDs {
```
Comment (knqyf263) [inline pkg/fanal/artifact/image/image.go:163]: I think we should compare `len(diffIDs)` and `len(createdBy)`. They should be the same. If they don't match, we should warn it. It is probably our bug.

File: pkg/report/table/secret_test.go [hunk 4]
```diff
@@ -115,7 +115,9 @@ HIGH: category (rule-id)
 ════════════════════════════════════════
 this is a title
 ────────────────────────────────────────
- my-file:3-4 (deleted in the intermediate layer)
+ my-file:3-4
+────────────────────────────────────────
+ This secret is added in 'COPY my-file my-file'
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/report/table/secret_test.go]: Done

File: pkg/fanal/types/secret.go [hunk 5]
```diff
@@ -16,6 +16,6 @@ type SecretFinding struct {
 	EndLine   int
 	Code      Code
 	Match     string
-	Deleted   bool
 	Layer     Layer `json:",omitempty"`
+	CreatedBy string
```
[the remark this answers is not in the document]
Comment (DmitriyLewen) [reply pkg/fanal/types/secret.go]: DOne

File: pkg/fanal/artifact/image/image.go (hunk 1 above)
Comment (DmitriyLewen) [reply pkg/fanal/artifact/image/image.go]: Done

**Review by ankk13 — COMMENTED** [review]
The deleted secret can be removed from secret map, if respective OpaqueDirs/Whiteoutfiles are found.

**Comment by DmitriyLewen** [comment]
Hello @ankk13 
You want to remove secret from secretMap, if it was removed in upper layer, right?

if so - we can't do that, because image still contain this secret.
It is not safely. User needs to remove secret from layer where this secret was added

**Comment by ankk13** [comment]
> ove secret from secretMap, if it was removed in upper layer, right?
> 
> if so - we can't do that, because image still contain this secret.

Yes, even if image still contain layer. I believe it is correct to tag secret as "delete in intermediate layer", as we are not commenting that secret is deleted. Let me know if you feel otherwise

**Review by ankk13 — APPROVED** [review]

**Review by tonaim — APPROVED** [review]

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**